src/common/io/http/request.c

```diff
@@ -302,6 +344,153 @@ httpRequestError(const HttpRequest *const this, HttpResponse *const response)
+        strCatZ(request, HTTP_HEADER_CONTENT_TRANSFER_ENCODING ":" HTTP_HEADER_CONTENT_TRANSFER_ENCODING_BINARY "\r\n");
+        strCatFmt(request, HTTP_HEADER_CONTENT_ID ":%s\r\n\r\n", strZ(contentId));
+        strCat(request, httpRequestFmt(verb, path, param.query, param.header, false));
+
+        MEM_CONTEXT_OBJ_BEGIN(this->contentList)
+        {
+            // Add content
+            Buffer *const content = bufNew(strSize(request) + (param.content == NULL ? 0 : bufUsed(param.content)));
+
+            bufCat(content, BUFSTR(request));
+            bufCat(content, param.content);
+
+            // Find a boundary that is not used in the content
+            while (bufFindP(content, this->boundaryRaw) != NULL)
+            {
+                if (bufUsed(this->boundaryRaw) == HTTP_MULTIPART_BOUNDARY_INIT_SIZE + HTTP_MULTIPART_BOUNDARY_EXTRA_SIZE)
+                    THROW(AssertError, "unable to construct unique boundary");
+
+                bufCatC(
+                    this->boundaryRaw, (unsigned char *)HTTP_MULTIPART_BOUNDARY_EXTRA,
+                    bufUsed(this->boundaryRaw) - HTTP_MULTIPART_BOUNDARY_INIT_SIZE, HTTP_MULTIPART_BOUNDARY_NEXT);
+            }
+
+            // Add to list
+            lstAdd(this->contentList, &content);
+            this->contentSize += bufUsed(content);
+        }
+        MEM_CONTEXT_OBJ_END();
+    }
+    MEM_CONTEXT_TEMP_END();
+
+    FUNCTION_TEST_RETURN_VOID();
+}
+
+/**********************************************************************************************************************************/
+#define HTTP_MULTIPART_BOUNDARY_PRE_SIZE                            (sizeof(HTTP_MULTIPART_BOUNDARY_PRE) - 1)
+#define HTTP_MULTIPART_BOUNDARY_POST_SIZE                           (sizeof(HTTP_MULTIPART_BOUNDARY_POST) - 1)
+
+FN_EXTERN Buffer *
+httpRequestMultiContent(HttpRequestMulti *const this)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(HTTP_REQUEST_MULTI, this);
+    FUNCTION_TEST_END();
+
+    ASSERT(this != NULL);
+    ASSERT(!lstEmpty(this->contentList));
+
+    const size_t boundarySize = bufUsed(this->boundaryRaw) + HTTP_MULTIPART_BOUNDARY_PRE_SIZE + HTTP_MULTIPART_BOUNDARY_POST_SIZE;
+    Buffer *const result = bufNew(this->contentSize + ((lstSize(this->contentList) + 1) * boundarySize));
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        // Generate boundary
+        Buffer *const boundary = bufNew(boundarySize);
+        bufCat(boundary, BUFSTRDEF(HTTP_MULTIPART_BOUNDARY_PRE));
+        bufCat(boundary, this->boundaryRaw);
+        bufCat(boundary, BUFSTRDEF(HTTP_MULTIPART_BOUNDARY_POST));
+
+        // Add first boundary
+        bufCat(result, boundary);
+
+        // Add content and boundaries
+        for (unsigned int contentIdx = 0; contentIdx < lstSize(this->contentList); contentIdx++)
+        {
+            bufCat(result, *(Buffer **)lstGet(this->contentList, contentIdx));
+            bufCat(result, boundary);
+        }
```

Comment:
Hey - I may be looking at the wrong reference(s), but examples that I've seen seem to have a two dash trailer on the last boundary entry in multicontent batch requests/responses?

Found this https://www.ietf.org/rfc/rfc2046.txt,
Page 19 paragraph 4
   The boundary delimiter line following the last body part is a
   distinguished delimiter that indicates that no further body parts
   will follow.  Such a delimiter line is identical to the previous
   delimiter lines, with the addition of two more hyphens after the
   boundary parameter value.
     --gc0pJq0M:08jU534c0p--

Reply:
No, I think you have it right and not sure how I missed it. Add in 3b2d778849.

I've decided to leave this check out of the parsing side, though, since GCS clearly does. There are other ways for us to know if the response is complete, i.e. content-length or content-encoding. My guess is this is a holdover from the HTTP 1.0 days when EOF was frequently used to denote end of content.